lib/pool.js

```diff
@@ -138,8 +138,9 @@ function getConnection(a1, a2) {
   var options = {};
 
   nodbUtil.assert(arguments.length >= 1 && arguments.length <= 2, 'NJS-009');
-  nodbUtil.assert(typeof getConnectionCb === 'function', 'NJS-006',
-      arguments.length);
+  nodbUtil.assert(typeof getConnectionCb === 'function', 'NJS-006', arguments.length);
+  nodbUtil.assert(self.closing === false, 'NJS-064', self.poolAlias);
```

Comment:
What about returning this in the callback?

```
-  nodbUtil.assert(self.closing === false, 'NJS-064', self.poolAlias);
-
+  if (self.closing === true) {
+    getConnectionCb(new Error(nodbUtil.getErrorMessage('NJS-064', self.poolAlias)));
+    return;
+  }
```

Reply:
The question is then, should there be a second message printed when the pool has finally closed?  The self.closing attribute would need to have 3 states.

Reply:
can we have a closing and hasClosed property? due to the nature of the closing being a boolean I would like to have a new property in this case.

Reply:
@danilohgds I know @anthony-tuininga is keen to add a check on connectionsOut at connection release which may change the flags a bit; I'll let him add his proposal.

Reply:
My suggestion was to store "_closeCb" (callback function) on the pool instead of "closing" (boolean). That would allow, as Chris mentioned, the ability to check connectionsOut at connection release. If the connectionsOut after release is zero, the close callback can be invoked immediately (but first that attribute would be reset to undefined so that when the timeout expires it won't call it a second time). This would permit well-behaved applications with reasonably short connection lifetimes to shut down quickly without waiting for the drain time to expire.

My other suggestion was to change the API to use an options object instead of a simple integer. The options object could contain one or both of these two keys:

- force: forces the pool to be closed even if connections are out (default false)
- drainTime: specifies the maximum amount of time the pool will wait for connections to be released back to the pool (default 0 seconds)

This provides the current behaviour (immediate close with a failure if any connections are out) and also provides a means to control both options. Currently if any value is specified for drainTime then it is assumed that forcing the close at the end of the drainTime value is desirable, and if no value is specified, then forcing is not desired. Its a bit awkward, but workable. Thoughts?

Reply:
@anthony-tuininga I think I understand your suggestion now. So essentially, the driver would be able to intervene and invoke the call back when the connectionsOut reaches 0 and save time.

As for the second suggestion, this would essentially mean that I can set up a drain time delay for both force and default option correct? I'm trying to see situations in which setting a drain time without the force would be used, since it might eventually fail if there still connections, maybe a routine that catches such error and schedules the closing again after X time? Might be beneficial I guess.

Reply:
I'm fine with the first suggestion (it completes https://github.com/oracle/node-oracledb/issues/836#issuecomment-400912975) but I'm not worried if it gets deferred.

I don't see a strong reason to have a drainTime for a non-force close.  Having what is effectively 'named' parameters does add flexibility for future OCI session pool enhancements, but I'm not sure it's worth it.  I'd be happy to go with the implemented 'less typing' API.

Reply:
I changed some of the code around for proper error messaging like you suggested, also cleaned up some variables and comments.

I will see how to make the ._close function stored and callable later during the connection releases.

Reply:
@danilohgds excellent.  Ping us if you want us to take over.